OpenRA.Game/Network/Connection.cs

```diff
@@ -87,22 +90,18 @@ protected virtual void Send(byte[] packet)
 
 		protected void AddPacket(ReceivedPacket packet)
 		{
-			lock (receivedPackets)
-				receivedPackets.Add(packet);
+			receivedPackets.Enqueue(packet);
 		}
 
-		public virtual void Receive(Action<int, byte[]> packetFn)
+		public virtual void Receive(Action<int, byte[], int> packetFn)
 		{
-			ReceivedPacket[] packets;
-			lock (receivedPackets)
-			{
-				packets = receivedPackets.ToArray();
-				receivedPackets.Clear();
-			}
+			var packets = new List<ReceivedPacket>(receivedPackets.Count);
+			while (receivedPackets.TryDequeue(out var received))
+				packets.Add(received);
 
 			foreach (var p in packets)
 			{
-				packetFn(p.FromClient, p.Data);
+				packetFn(p.FromClient, p.Data, p.Timestep);
```

Comment:
https://github.com/OpenRA/OpenRA/pull/18783#discussion_r606718799 still valid?

Reply:
Will need to take another stab at understanding the full picture before commenting about Timestep vs Data, but wouldnt it make sense in any case for `packetFn` to take the `ReceivedPacket` directly?

Reply:
Moving timestep into data should work, but we'd also have to account for this in the old code then. But I'm not sure if we should do that. Data is sent over the network, timestep isn't(?).

Reply:
#19632 reworks this to be a lot cleaner. After a nasty rebase, this PR will be able to use a new Timestep order type and call a handler on `OrderManager` directly.